### PR TITLE
[ACC-22] add tunable client props to pika

### PR DESCRIPTION
**Pull request by mffrench**
Hello,

I added a little functionality which help you to define the application client properties. eg : 

```
import pika

client_properties = {
    'product': 'my product',
    'information': 'my information',
    'my_other_props': 'my other property',
}

credentials = pika.PlainCredentials('guest', 'guest')
parameters = pika.ConnectionParameters("localhost", 5672, '/',
                                       credentials=credentials, client_props=client_properties)
connection = pika.BlockingConnection(parameters)
```

You'll be then able to differentiate python connections from each other more easily...

Cheers,

Mathilde
screen shot 2015-05-14 at 12 26 26

### CODE REVIEW

**Comment by mffrench** [comment]
Hello guys,

please tell me if there is any other blocking points for that pull request...

Cheers,

Mathilde

**Comment by gmr** [comment]
I've not had a chance to review it but from your screenshot it looks like it nukes property that defines it as pika. I'm not a fan of that. I'm not opposed to allowing the addition of new properties, though I don't think it should be in the constructor. Most static analysis and lint tools complain about the argument counts already.

**Comment by mffrench** [comment]
About nuking properties from pika : I reproduced behaviour of the Java RabbitMQ lib. 

As a former MoM OPS I think the product connecting to RabbitMQ is not the libraries (Java, Python or whatever) but the application calling this lib as some time you can have a lot of connection from many different apps into the same MoM and knowing the apps connecting on it is much more relevant than knowing the lib used by the app... So I think the java lib on this point has the good behaviour by allowing client properties redefinition.
 
If you want to keep Pika reference on the client properties I personally would prefer to add a new client property like library=Pika 0.9.1.

About the contructor : do you prefer to add client_props setter/getter instead of passing through the ConnectionParameters constructor ? Or do you have other solution in mind ?

Cheers

**Comment by mffrench** [comment]
Dear Pika project collaborators,

As there may be here a disagreement around RabbitMQ client properties usage I would like to know the final status of this pull req (rejected or accepted) ? If rejected I will have to maintain a specific fork to fit my needs. 

Best regards,

Mathilde

**Comment by wjps** [comment]
FWIW I'd like to see this or something similar merged, it would be useful
to us. I also think the product and version fields should be settable by
the application, I'd be inclined to put 'pika' and it's version in a new
field.

**Comment by mffrench** [comment]
Dear Pika contributors,

currently waiting the status of this request. Meanwhile I pushed my own distribs adding this functionality to Pika 0.9.14 : 

for master : https://github.com/echinopsii/net.echinopsii.3rdparty.pika/tree/tune_client_props

for last release on python 2 (0.9.14) : https://github.com/echinopsii/net.echinopsii.3rdparty.pika/tree/tune_client_props_0.9.14
pip install epika-python2

for last release on python 3 (0.9.14) : 
https://github.com/echinopsii/net.echinopsii.3rdparty.pika/tree/tune_client_props_p3_0.9.14
pip3 install epika-python3 

Let me know when this feature will be merged or if you plan to develop something similar, I will then stop maintaining these branches.

Best regards,

Mathilde

**Comment by markunsworth** [comment]
I think this is a useful option. I do agree, as @mffrench suggests, that the application is the client here, but it does make sense to hold environment properties (python version, pika version) alongside application specific properties.

The AMQP spec defines:

> The properties SHOULD contain at least these fields: "product", giving the name of the client product, "version", giving the name of the client version, "platform", giving the name of the operating system, "copyright", if appropriate, and "information", giving other general information.


What about defining a structure that meets the specification for client properties but that allows application defined settings as well as pika defined ones as well. 

    {
        "product": "Pika",
        "version": "0.10.0.b",
        "platform": "Python 2.7.10",
        "copyright": "",
        "information": {
            "pika_version": "0.10.0.b",
            "python_version": "2.7.10",
            "process_id": 237,
            "hostname": "my-server.eu.aws, 
            ...                            
        },
    }

By default we set the product, version and platform as we do currently and also set them in the information. If a client specifies the product, version or platform they should override the defaults. We then don't lose information, but it gives app developers flexibility

----


We've been setting the hostname by default on all connections, just added as #628 as it's helps with debugging as we run multiple instances of each service in our system.

image

**Comment by mffrench** [comment]
Dear Pika contributors,

since pika 0.9.14 we do not have any clear answer regarding this pull req. I'm looking to your last version and asking my self if I will report my changes in a separate packaging like I did with pika 0.9.14 ( https://github.com/echinopsii/net.echinopsii.3rdparty.pika/tree/tune_client_props).

So could you please tell us if you plan to merge this pull req and if yes when ?

Thank you

**Comment by barryib** [comment]
+1 for this pull request.

In our case, we wanted to use client properties for debug and operational purposes. In fact, our rabbitmq clusters are behind load balancers, so the rabbitmq admin console shows the load balancer's ip address for connections, instead of the original client ip. To solve that problem, we wanted to add the client ip address or hostname within the client properties. 

For the **product** field in client properties, I think that it should refer to the application name instead of the used library. An additional filed can be added for library.

File: pika/connection.py [hunk 1]
```diff
@@ -319,6 +319,7 @@ class ConnectionParameters(Parameters):
     :param int|float socket_timeout: Use for high latency networks
     :param str locale: Set the locale value
     :param bool backpressure_detection: Toggle backpressure detection
+    :param json object client_props: setup your client properties (and overrides the default ones)
```
Comment (vitaly-krugl) [inline pika/connection.py]: Couple of problems here:

1. in python, it's a "dict", not "json object"
2. "json object" is two words; only one word types are allowed here per rst syntax.

File: pika/connection.py [hunk 2]
```diff
@@ -398,6 +401,7 @@ def __init__(self,
         if (backpressure_detection is not None and
             self._validate_backpressure(backpressure_detection)):
             self.backpressure_detection = backpressure_detection
+        self.client_props = client_props
 
 
 class URLParameters(Parameters):
```
Comment (vitaly-krugl) [inline pika/connection.py]: How would someone who needs URLParameters make use of this new feature?

File: pika/connection.py [hunk 3]
```diff
@@ -336,7 +337,8 @@ def __init__(self,
                  retry_delay=None,
                  socket_timeout=None,
                  locale=None,
-                 backpressure_detection=None):
+                 backpressure_detection=None,
+                 client_props=None):
```
Comment (vitaly-krugl) [inline pika/connection.py]: What about integration with the `Parameters` class, as is the case for all other args, including `DEFAULT_*` and `_validate_*`?

**Comment by vitaly-krugl** [comment]
What about tests?

**Comment by mffrench** [comment]
Don't have time to fulfill your complains now. If you're interested by this feature I'm sure you'll find the time to complete this PR the way you think it should be. 

Cheers

(FYI : next time I'm working on this subject - if this PR has not been merged - I will just report this PR on the last pika tag to update our current pika fork.)

**Comment by vitaly-krugl** [comment]
@mffrench - not so much a complaint, but rather peer code review feedback on items impacting quality and stability of the code base.

Best,
Vitaly

**Comment by mffrench** [comment]
Don't worry I was not upset.  You're right to push your comments for code quality/stability but I really don't have time to make changes now - even to look forward what should be done. 

Currently we're using echinopsii's pika branch which is working well for our current needs and I even don't know when I will upgrade it... 

So if you(*)'re interested on this functionality, feel free to complete this PR following the recommendations you made.

Cheers,

(*) 'you' means you (@vitaly-krugi) or anybody else interested on this functionality and wanting to go ahead on this topic...

**Comment by gmr** [comment]
Will address this when https://github.com/vitaly-krugl/pika/pull/4 is merged into master here.

**Pull request closed without merging**